_websiteutils/make_redirects_links.py

```diff
@@ -0,0 +1,242 @@
+                    ll = rec.sub(new.encode("utf-8"), line)
+                    _log.debug(f"new {line}->{ll}")
+                    fout.write(ll)
+                    found = True
+                elif b'<body>' in line and not newest:
+                    # add a warning right under:
+                    fout.write(line)
+                    line = next(fin)
+                    if last == tocheck[0]:
+                        new = warn_banner_exists.format(
+                            version=fullname.parts[0],
+                            url=newcanon)
+                    else:
+                        new = warn_banner_old.format(version=fullname.parts[0])
+                    fout.write(new.encode("utf-8"))
+                    if b'<div id="olddocs-message">' not in line:
+                        # write the line out if it wasn't an olddocs-message:
+                        fout.write(line)
+
+                else:
+                    fout.write(line)
+
+    shutil.move(fout.name, fullname)
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--np", type=int, help="Number of processors to use")
+    parser.add_argument("--no-canonicals", help="do not do canonical links",
+                        action="store_true")
+    parser.add_argument("--no-redirects", help="do not do redirects links",
+                        action="store_true")
+
+    args = parser.parse_args()
+    if args.np:
+        np = args.np
+    else:
+        np = None
+
+    # figure out the newest version and trim tocheck at the same time:
+    tocheck = tuple(p for p in tocheck if p.exists())
+    print(tocheck)
+
+    # html redirect or soft link most things in the top-level directory that
+    # are not other modules or versioned docs.
+    if not args.no_redirects:
+        for entry in os.scandir("."):
+            fullname = pathlib.Path(entry.name)
+            if fullname not in toignore:
+                if entry.is_dir():
+                    do_links(entry.name)
+                elif fullname.suffix == ".html":
+                    last = findlast(fullname, tocheck)
+                    _log.debug(f"Checking: {fullname} found {last}")
+                    if last is not None:
+                        fullname.unlink()
+                        _log.info(f"Rewriting HTML: {fullname} in {last}")
+                        with fullname.open("w") as fout:
+                            oldname = last / fullname
+                            st = html_redirect.format(newurl=oldname,
+                                                      canonical=oldname)
+                            fout.write(st)
+        _log.info("Done links and redirects")
+
+    # change the canonical url for all html to the newest version in the docs:
+    if not args.no_canonicals:
+        if np is not None:
+            with multiprocessing.Pool(np) as pool:
+                pool.map(do_canonicals, tocheck[1:])
```

Comment:
This option now fails: 
```
Traceback (most recent call last):
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/Users/jklymak/matplotlib.github.com/_websiteutils/make_redirects_links.py", line 142, in do_canonicals
    last = findlast(basename, tocheck)
TypeError: unhashable type: 'list'
```

Reply:
Perhaps just remove it?

Reply:
Ah, `map` must convert it from a tuple to a list.

Reply:
so  its the cache that is causing the problem?  Happy to remove my optimization in favour of your optimization ;-)

Reply:
Actually no, this works fine for me; do you have some stashed changes? `tocheck` should be a tuple after df8ed616dd53bcf4393d3c10b672eb622d0d468d (which was before 6fcc3b744733fb9081825a96731fa5d3b42ce812).

Reply:
That aside, I think this is working great....